src/java/org/apache/cassandra/streaming/StreamSession.java

```diff
@@ -156,18 +166,57 @@
     private final NettyStreamingMessageSender messageSender;
     private final ConcurrentMap<ChannelId, Channel> incomingChannels = new ConcurrentHashMap<>();
 
-    private final AtomicBoolean isAborted = new AtomicBoolean(false);
+    // "maybeCompleted()" should be executed at most once. Because it can be executed asynchronously by IO
+    // threads(serialization/deserialization) and stream messaging processing thread, causing connection closed before
+    // receiving peer's CompleteMessage.
+    private final AtomicBoolean maybeCompleted = new AtomicBoolean(false);
+    private final AtomicBoolean isClosed = new AtomicBoolean(false);
+    private Future closeFuture;
+
     private final UUID pendingRepair;
     private final PreviewKind previewKind;
 
+    /**
+     * State Transition:
+     *
+     * <pre>
+     *
+     *  +------------------+----------> FAILED <------------------+
+     *  |                  |              ^                       |
+     *  |                  |              |                       |
+     *  INITIALIZED --> PREPARING --> STREAMING ----------> WAIT_COMPLETE ----> COMPLETED
+     *  |                  |                                      ^                 ^
+     *  |                  |         if preview                   |                 |
+     *  |                  +--------------------------------------+                 |
+     *  |               nothing to request or to transfer                           |
+     *  +---------------------------------------------------------------------------+
+     *                  nothing to request or to transfer
+     *
+     *  </pre>
+     */
     public enum State
     {
-        INITIALIZED,
-        PREPARING,
-        STREAMING,
-        WAIT_COMPLETE,
-        COMPLETE,
-        FAILED,
+        INITIALIZED(false),
+        PREPARING(false),
+        STREAMING(false),
+        WAIT_COMPLETE(false),
+        COMPLETE(true),
+        FAILED(true);
+
+        private final boolean finalState;
+
+        State(boolean finalState)
+        {
+            this.finalState = finalState;
+        }
+
+        /**
+         * @return true if current statu is final and cannot be change to other state.
```

Comment:
nit: `statu` -> `status`, `change` -> `changed`.

Reply:
Also, this doesn't seem to be actually enforced? That is, we still allow to go from a final state to a different state.

Reply:
updated the javadoc.. State transition is not enforced, leaving it to [C-14754](https://issues.apache.org/jira/browse/CASSANDRA-14754)